core/lib/multivm/src/tracers/call_tracer/era_vm/mod.rs

```diff
@@ -0,0 +1,188 @@
+                // We use parent gas for properly calculating gas used in the trace.
+                let current_ergs = execution.gas_left().unwrap();
+                let parent_gas = execution
+                    .running_contexts
+                    .last()
+                    .map(|call| call.frame.gas_left.0.saturating_add(current_ergs))
+                    .unwrap_or(current_ergs) as u64;
+
+                // we need to do this cast because `Call` uses another library
+                let far_call_variant = match far_call as u8 {
+                    0 => FarCallOpcode::Normal,
+                    1 => FarCallOpcode::Delegate,
+                    2 => FarCallOpcode::Mimic,
+                    _ => FarCallOpcode::Normal, // Should never happen
+                };
+
+                let mut current_call = Call {
+                    r#type: CallType::Call(far_call_variant),
+                    gas: 0,
+                    parent_gas: parent_gas as u64,
+                    ..Default::default()
+                };
+
+                self.handle_far_call_op_code_era(execution, &mut current_call);
+                self.push_call_and_update_stats(current_call, 0);
+            }
+            Variant::Ret(ret_code) => {
+                self.handle_ret_op_code_era(execution, ret_code);
+            }
+            _ => {}
+        };
+    }
+}
+
+impl<S: ReadStorage> VmTracer<S> for CallTracer {
+    fn after_bootloader_execution(&mut self, _state: &mut crate::era_vm::vm::Vm<S>) {
+        self.store_result();
+    }
+}
+
+impl CallTracer {
+    fn handle_far_call_op_code_era(&mut self, execution: &Execution, current_call: &mut Call) {
+        // since this is a far_call, the current_context represents the current frame
+        let current = execution.current_context().unwrap();
+        // All calls from the actual users are mimic calls,
+        // so we need to check that the previous call was to the deployer.
+        // Actually it's a call of the constructor.
+        // And at this stage caller is user and callee is deployed contract.
+        let call_type = if let CallType::Call(far_call) = current_call.r#type {
+            if matches!(far_call, FarCallOpcode::Mimic) {
+                let previous_caller = execution
+                    .running_contexts
+                    .first()
+                    .map(|call| call.caller)
+                    // Actually it's safe to just unwrap here, because we have at least one call in the stack
+                    // But i want to be sure that we will not have any problems in the future
+                    .unwrap_or(current.caller);
+                if previous_caller == CONTRACT_DEPLOYER_ADDRESS {
+                    CallType::Create
+                } else {
+                    CallType::Call(far_call)
+                }
+            } else {
+                CallType::Call(far_call)
+            }
+        } else {
+            return;
+        };
```

Comment:
You can reduce nesting a bit by using `let <destructuring pattern> else` here:
```suggestion
        let CallType::Call(far_call) = current_call.r#type else {
            return;
        }
        if matches!(far_call, FarCallOpcode::Mimic) {
            let previous_caller = execution
                .running_contexts
                .first()
                .map(|call| call.caller)
            // Actually it's safe to just unwrap here, because we have at least one call in the stack
            // But i want to be sure that we will not have any problems in the future
                .unwrap_or(current.caller);
            if previous_caller == CONTRACT_DEPLOYER_ADDRESS {
                CallType::Create
            } else {
                CallType::Call(far_call)
            }
        } else {
            CallType::Call(far_call)
        };
```

Reply:
The thing is that this code is exactly as is for the other vms, so I think it is better to leave it as is. The only change was the one from the unreachable, so we don't stop execution due to a tracer (even though in the other vms they still have it)